Google.Api.Gax.Grpc/Rest/HttpRulePathPattern.cs

```diff
@@ -48,19 +48,31 @@ internal static HttpRulePathPattern Parse(string pattern, MessageDescriptor requ
                     endOfName = match.Value.Length - 1;
                 }
                 string propertyPath = match.Value.Substring(1, endOfName - 1);
-                segments.Add(CreatePropertyAccessor(requestDescriptor, propertyPath));
+                segments.Add(CreatePathPathPatternSegment(requestDescriptor, propertyPath));
                 lastEnd = match.Index + match.Length;
             }
             if (lastEnd != pattern.Length)
             {
                 string literal = pattern.Substring(lastEnd);
-                segments.Add(message => literal);
+                segments.Add(HttpRulePathPatternSegment.CreateLiteral(literal));
             }
             return new HttpRulePathPattern(segments);
         }
 
         // TODO: do we need to perform URL encoding of anything?
-        internal string Format(IMessage message) => string.Join("", _segments.Select(segment => segment(message)));
+        internal string Format(IMessage message) => string.Join("", _segments.Select(segment => segment.Format(message)));
+
+        internal List<string> TopLevelFieldNames => _segments.Where(segment => segment.TopLevelFieldName != null)
+            .Select(s => s.TopLevelFieldName).ToList();
```

Comment:
I'd *suggest* reformatting as this:

```suggestion
        internal List<string> TopLevelFieldNames => _segments
            .Where(segment => segment.TopLevelFieldName != null)
            .Select(s => s.TopLevelFieldName)
            .ToList();
```

Or possibly even put `_segments` on the second line and indent the query calls further. I'm not bothered which.

Basically if a LINQ query is too long for a single line, we tend to prefer "one operator per line".

Reply:
👍